From: Ilya Dryomov <idryomov@gmail.com>
Date: Mon, 12 Sep 2016 18:59:42 +0200
Subject: rbd: change rbd_obj_request_submit() signature
Git-commit: 980917fc6ec94cb614fd79e6a124689e700f9d97
Patch-mainline: v4.9-rc1
References: FATE#322288

- osdc parameter is useless
- starting with commit 5aea3dcd5021 ("libceph: a major OSD client
  update"), ceph_osdc_start_request() always returns success

Signed-off-by: Ilya Dryomov <idryomov@gmail.com>
Reviewed-by: Alex Elder <elder@linaro.org>
Reviewed-by: David Disseldorp <ddiss@suse.de>
Acked-by: Luis Henriques <lhenriques@suse.com>
[luis: modified several other callers: rbd_img_obj_creatrunc_submit,
 rbd_obj_notify_ack_sync, rbd_obj_watch_request_helper, rbd_dev_setxattr,
 rbd_dev_cmpsetxattr and rbd_dev_getxattr.]
---
 drivers/block/rbd.c |  105 ++++++++++++++--------------------------------------
 1 file changed, 29 insertions(+), 76 deletions(-)

--- a/drivers/block/rbd.c
+++ b/drivers/block/rbd.c
@@ -1470,11 +1470,12 @@ static bool obj_request_type_valid(enum
 	}
 }
 
-static int rbd_obj_request_submit(struct ceph_osd_client *osdc,
-				struct rbd_obj_request *obj_request)
+static void rbd_obj_request_submit(struct rbd_obj_request *obj_request)
 {
-	dout("%s %p\n", __func__, obj_request);
-	return ceph_osdc_start_request(osdc, obj_request->osd_req, false);
+	struct ceph_osd_request *osd_req = obj_request->osd_req;
+
+	dout("%s %p osd_req %p\n", __func__, obj_request, osd_req);
+	ceph_osdc_start_request(osd_req->r_osdc, osd_req, false);
 }
 
 static void rbd_obj_request_end(struct rbd_obj_request *obj_request)
@@ -2783,7 +2784,6 @@ rbd_img_obj_parent_read_full_callback(st
 {
 	struct rbd_obj_request *orig_request;
 	struct ceph_osd_request *osd_req;
-	struct ceph_osd_client *osdc;
 	struct rbd_device *rbd_dev;
 	struct page **pages;
 	enum obj_operation_type op_type;
@@ -2820,13 +2820,9 @@ rbd_img_obj_parent_read_full_callback(st
 	 * and re-submit the original write request.
 	 */
 	if (!rbd_dev->parent_overlap) {
-		struct ceph_osd_client *osdc;
-
 		ceph_release_page_vector(pages, page_count);
-		osdc = &rbd_dev->rbd_client->client->osdc;
-		img_result = rbd_obj_request_submit(osdc, orig_request);
-		if (!img_result)
-			return;
+		rbd_obj_request_submit(orig_request);
+		return;
 	}
 
 	if (img_result)
@@ -2860,10 +2856,9 @@ rbd_img_obj_parent_read_full_callback(st
 
 	/* All set, send it off. */
 
-	osdc = &rbd_dev->rbd_client->client->osdc;
-	img_result = rbd_obj_request_submit(osdc, orig_request);
-	if (!img_result)
-		return;
+	rbd_obj_request_submit(orig_request);
+	return;
+
 out_err:
 	/* Record the error code and complete the request */
 
@@ -2997,17 +2992,13 @@ static void rbd_img_obj_exists_callback(
 
 	/*
 	 * If the overlap has become 0 (most likely because the
-	 * image has been flattened) we need to free the pages
-	 * and re-submit the original write request.
+	 * image has been flattened) we need to re-submit the
+	 * original request.
 	 */
 	rbd_dev = orig_request->img_request->rbd_dev;
 	if (!rbd_dev->parent_overlap) {
-		struct ceph_osd_client *osdc;
-
-		osdc = &rbd_dev->rbd_client->client->osdc;
-		result = rbd_obj_request_submit(osdc, orig_request);
-		if (!result)
-			return;
+		rbd_obj_request_submit(orig_request);
+		return;
 	}
 
 	/*
@@ -3039,7 +3030,6 @@ static int rbd_img_obj_exists_submit(str
 {
 	struct rbd_obj_request *stat_request;
 	struct rbd_device *rbd_dev;
-	struct ceph_osd_client *osdc;
 	struct page **pages = NULL;
 	u32 page_count;
 	size_t size;
@@ -3083,8 +3073,9 @@ static int rbd_img_obj_exists_submit(str
 					false, false);
 	rbd_osd_req_format_read(stat_request);
 
-	osdc = &rbd_dev->rbd_client->client->osdc;
-	ret = rbd_obj_request_submit(osdc, stat_request);
+	rbd_obj_request_submit(stat_request);
+	return 0;
+
 out:
 	if (ret)
 		rbd_obj_request_put(obj_request);
@@ -3197,15 +3188,10 @@ rbd_img_obj_creatrunc_submit(struct rbd_
 	/* rbd_osd_req_format_write() using snapc from img_request */
 	creatrunc_req->osd_req->r_mtime = CURRENT_TIME;
 
-	ret = rbd_obj_request_submit(osdc, creatrunc_req);
-	if (ret) {
-		goto fail_submit;
-	}
+	rbd_obj_request_submit(creatrunc_req);
 
 	return 0;
 
-fail_submit:
-	rbd_obj_request_put(obj_request);
 fail_creatrunc_request:
 	creatrunc_req->img_request = NULL;
 	rbd_obj_request_put(creatrunc_req);
@@ -3260,13 +3246,8 @@ static bool img_obj_request_simple(struc
 static int rbd_img_obj_request_submit(struct rbd_obj_request *obj_request)
 {
 	if (img_obj_request_simple(obj_request)) {
-		struct rbd_device *rbd_dev;
-		struct ceph_osd_client *osdc;
-
-		rbd_dev = obj_request->img_request->rbd_dev;
-		osdc = &rbd_dev->rbd_client->client->osdc;
-
-		return rbd_obj_request_submit(osdc, obj_request);
+		rbd_obj_request_submit(obj_request);
+		return 0;
 	}
 
 	/*
@@ -3330,12 +3311,8 @@ static void rbd_img_parent_read_callback
 	rbd_assert(obj_request->img_request);
 	rbd_dev = obj_request->img_request->rbd_dev;
 	if (!rbd_dev->parent_overlap) {
-		struct ceph_osd_client *osdc;
-
-		osdc = &rbd_dev->rbd_client->client->osdc;
-		img_result = rbd_obj_request_submit(osdc, obj_request);
-		if (!img_result)
-			return;
+		rbd_obj_request_submit(obj_request);
+		return;
 	}
 
 	obj_request->result = img_result;
@@ -3415,7 +3392,6 @@ out_err:
 static int rbd_obj_notify_ack_sync(struct rbd_device *rbd_dev, u64 notify_id)
 {
 	struct rbd_obj_request *obj_request;
-	struct ceph_osd_client *osdc = &rbd_dev->rbd_client->client->osdc;
 	int ret;
 
 	obj_request = rbd_obj_request_create(rbd_dev->header_oid.name, 0, 0,
@@ -3433,9 +3409,7 @@ static int rbd_obj_notify_ack_sync(struc
 			      CEPH_OSD_OP_NOTIFY_ACK, 0, notify_id);
 	rbd_osd_req_format_read(obj_request);
 
-	ret = rbd_obj_request_submit(osdc, obj_request);
-	if (ret)
-		goto out;
+	rbd_obj_request_submit(obj_request);
 	ret = rbd_obj_request_wait(obj_request);
 out:
 	rbd_obj_request_put(obj_request);
@@ -3502,10 +3476,7 @@ static struct rbd_obj_request *rbd_obj_w
 	    watch_opcode == CEPH_OSD_WATCH_OP_WATCH)
 		ceph_osdc_set_request_linger(osdc, obj_request->osd_req);
 
-	ret = rbd_obj_request_submit(osdc, obj_request);
-	if (ret)
-		goto out;
-
+	rbd_obj_request_submit(obj_request);
 	ret = rbd_obj_request_wait_timeout(obj_request, opts->mount_timeout);
 	if (ret)
 		goto out;
@@ -3609,7 +3580,6 @@ static int rbd_obj_method_sync(struct rb
 			     void *inbound,
 			     size_t inbound_size)
 {
-	struct ceph_osd_client *osdc = &rbd_dev->rbd_client->client->osdc;
 	struct rbd_obj_request *obj_request;
 	struct page **pages;
 	u32 page_count;
@@ -3660,9 +3630,7 @@ static int rbd_obj_method_sync(struct rb
 					0, false, false);
 	rbd_osd_req_format_read(obj_request);
 
-	ret = rbd_obj_request_submit(osdc, obj_request);
-	if (ret)
-		goto out;
+	rbd_obj_request_submit(obj_request);
 	ret = rbd_obj_request_wait(obj_request);
 	if (ret)
 		goto out;
@@ -3831,7 +3799,6 @@ static int rbd_obj_read_sync(struct rbd_
 				u64 offset, u64 length, void *buf)
 
 {
-	struct ceph_osd_client *osdc = &rbd_dev->rbd_client->client->osdc;
 	struct rbd_obj_request *obj_request;
 	struct page **pages = NULL;
 	u32 page_count;
@@ -3866,9 +3833,7 @@ static int rbd_obj_read_sync(struct rbd_
 					false, false);
 	rbd_osd_req_format_read(obj_request);
 
-	ret = rbd_obj_request_submit(osdc, obj_request);
-	if (ret)
-		goto out;
+	rbd_obj_request_submit(obj_request);
 	ret = rbd_obj_request_wait(obj_request);
 	if (ret)
 		goto out;
@@ -4312,7 +4277,6 @@ static ssize_t rbd_image_refresh(struct
 int rbd_dev_setxattr(struct rbd_device *rbd_dev, char *key, void *val,
 		     int val_len)
 {
-	struct ceph_osd_client *osdc = &rbd_dev->rbd_client->client->osdc;
 	struct rbd_obj_request *obj_request;
 	int ret;
 
@@ -4335,10 +4299,7 @@ int rbd_dev_setxattr(struct rbd_device *
 
 	rbd_osd_req_format_write(obj_request);
 
-	ret = rbd_obj_request_submit(osdc, obj_request);
-	if (ret)
-		goto out;
-
+	rbd_obj_request_submit(obj_request);
 	ret = rbd_obj_request_wait(obj_request);
 	if (ret)
 		goto out;
@@ -4359,7 +4320,6 @@ EXPORT_SYMBOL(rbd_dev_setxattr);
 int rbd_dev_cmpsetxattr(struct rbd_device *rbd_dev, char *key, void *oldval,
 			int oldval_len, void *newval, int newval_len)
 {
-	struct ceph_osd_client *osdc = &rbd_dev->rbd_client->client->osdc;
 	struct rbd_obj_request *obj_request;
 	int ret;
 
@@ -4392,10 +4352,7 @@ int rbd_dev_cmpsetxattr(struct rbd_devic
 
 	rbd_osd_req_format_write(obj_request);
 
-	ret = rbd_obj_request_submit(osdc, obj_request);
-	if (ret)
-		goto out;
-
+	rbd_obj_request_submit(obj_request);
 	ret = rbd_obj_request_wait(obj_request);
 	if (ret)
 		goto out;
@@ -4416,7 +4373,6 @@ EXPORT_SYMBOL(rbd_dev_cmpsetxattr);
 int rbd_dev_getxattr(struct rbd_device *rbd_dev, char *key, int max_val_len,
 		     void **_val, int *val_len)
 {
-	struct ceph_osd_client *osdc = &rbd_dev->rbd_client->client->osdc;
 	struct rbd_obj_request *obj_request;
 	int page_count;
 	struct page **pages = NULL;
@@ -4459,10 +4415,7 @@ int rbd_dev_getxattr(struct rbd_device *
 					     0, false, false);
 	rbd_osd_req_format_read(obj_request);
 
-	ret = rbd_obj_request_submit(osdc, obj_request);
-	if (ret)
-		goto out;
-
+	rbd_obj_request_submit(obj_request);
 	ret = rbd_obj_request_wait(obj_request);
 	if (ret)
 		goto out;
